src/server/scripts/EasternKingdoms/MagistersTerrace/boss_vexallus.cpp

```diff
@@ -15,124 +15,239 @@
+        }
+    }
+
+    // Only take damage from players targeting this add
+    void DamageTaken(Unit* attacker, uint32& damage, DamageEffectType, SpellSchoolMask) override
+    {
+        if (!attacker || !attacker->IsPlayer() || attacker->GetVictim() != me)
+            damage = 0;
+    }
+
+    // Apply feedback effect on death
+    void JustDied(Unit* killer) override
+    {
+        if (killer)
+            killer->CastSpell(killer, SPELL_ENERGY_FEEDBACK, true, nullptr, nullptr, me->GetGUID());
+    }
+
+    // Override to prevent actual melee attacks while still allowing movement
+    void AttackStart(Unit* victim) override 
+    {
+        if (victim)
+        {
+            me->Attack(victim, false); // Set victim but don't allow melee
+            me->GetMotionMaster()->MoveChase(victim, 0.0f, 0.0f);
+        }
+    }
 };
 
+// Main boss AI implementation
 struct boss_vexallus : public BossAI
 {
-    boss_vexallus(Creature* creature) : BossAI(creature, DATA_VEXALLUS), _energyCooldown(false) { }
+    // Configuration constants
+    static constexpr std::array<uint8_t, 5> HEALTH_THRESHOLDS = {85, 70, 55, 40, 25}; // HP % for energy spawns
+    static constexpr Seconds ENERGY_COOLDOWN = 5s;    // Time between energy spawns
+    static constexpr Seconds ABILITY_TIMER = 8s;      // Basic ability cooldown
+    static constexpr uint8_t INITIAL_PURE_ENERGY = 5; // Starting energy count
+
+    explicit boss_vexallus(Creature* creature)
+        : BossAI(creature, DATA_VEXALLUS)
+        , _energyCooldown(false)
+        , _overloaded(false)
+        , _energyQueue(0)
+        , _pureEnergy(INITIAL_PURE_ENERGY)
+        , _thresholdsPassed({})
+    {
+    }
 
+    // Reset encounter state
     void Reset() override
     {
         _Reset();
         _energyCooldown = false;
+        _energyQueue = 0;
+        _overloaded = false;
+        _pureEnergy = INITIAL_PURE_ENERGY;
+        std::fill(_thresholdsPassed.begin(), _thresholdsPassed.end(), false);
 
-        ScheduleHealthCheckEvent({ 85, 70, 55, 40, 25 }, [&]
-        {
-            if (!_energyCooldown)
-            {
-                Talk(SAY_ENERGY);
-                Talk(EMOTE_DISCHARGE_ENERGY);
-
-                if (IsHeroic())
-                {
-                    DoCastSelf(SPELL_SUMMON_PURE_ENERGY_H1);
-                    DoCastSelf(SPELL_SUMMON_PURE_ENERGY_H2);
-                }
-                else
-                    DoCastSelf(SPELL_SUMMON_PURE_ENERGY_N);
-
-                _energyCooldown = true;
-                scheduler.Schedule(5s, [this](TaskContext)
-                {
-                    _energyCooldown = false;
-                });
-            }
-        });
+        ScheduleEnergyCheck();
+    }
 
-        ScheduleHealthCheckEvent(20, [&]
-        {
-            scheduler.CancelAll();
-            DoCastSelf(SPELL_OVERLOAD, true);
-        });
+    // Cleanup on death
+    void JustDied(Unit*) override
+    {
+        _JustDied();
+        if (instance)
```

Comment:
Boss scripts wont execute outside instances, there's no need to check for instance here